src/relax/transform/simplify_norm.cc

```diff
@@ -73,10 +74,60 @@ Expr SimplifyBatchNorm(const CallNode* call) {
   return out;
 }
 
-/*! \brief A mutator to simplify the normalization inference. */
-class NormInferenceSimplifier : public ExprMutator {
+Array<Optional<Expr>> SimplifyBatchNormTraining(const CallNode* call) {
+  auto attrs = call->attrs.as<BatchNormAttrs>();
+  ICHECK_NOTNULL(attrs);
+
+  Expr data = call->args[0];
+  TensorStructInfo sinfo = MatchTensorStructInfo(data);
+  Expr gamma = call->args[1];
+  Expr beta = call->args[2];
+
+  Array<Integer> reduce_axes;
+  for (int i = 0; i < sinfo->ndim; ++i) {
+    if (i != attrs->axis) {
+      reduce_axes.push_back(i);
+    }
+  }
+
+  Expr data_mean = mean(data, reduce_axes, false);
+  Expr data_mean_rs = ExpandToMatchInput(data_mean, sinfo->ndim, {attrs->axis});
+  Expr data_var = variance(data, reduce_axes, false);
+  Expr data_var_rs = ExpandToMatchInput(data_var, sinfo->ndim, {attrs->axis});
+
+  // output = (x - mean) / sqrt(var + epsilon) * gamma + beta
+  Expr epsilon = MakeConstantScalar(static_cast<float>(attrs->epsilon), sinfo->dtype);
+  Expr sqrt_var = sqrt(add(data_var_rs, epsilon));
+  Expr out = divide(subtract(data, data_mean_rs), sqrt_var);
+
+  if (attrs->scale) {
+    out = multiply(out, ExpandToMatchInput(gamma, sinfo->ndim, {attrs->axis}));
+  }
+  if (attrs->center) {
+    out = add(out, ExpandToMatchInput(beta, sinfo->ndim, {attrs->axis}));
+  }
+
+  Expr moving_mean = call->args[3];
+  Expr moving_var = call->args[4];
+  Expr momentum = MakeConstantScalar(static_cast<float>(attrs->momentum), sinfo->dtype);
+  Expr one_minus_mom = MakeConstantScalar(static_cast<float>(1 - attrs->momentum), sinfo->dtype);
+
+  return {
+      out,
+      add(multiply(one_minus_mom, moving_mean), multiply(momentum, data_mean)),
+      add(multiply(one_minus_mom, moving_var), multiply(momentum, data_var)),
+  };
+}
+
+static const char* kModeEval = "eval";
+static const char* kModeTraining = "training";
+
+/*! \brief A mutator to simplify the normalization. */
+class NormSimplifier : public ExprMutator {
  public:
-  static Expr Simplify(Expr expr) { return NormInferenceSimplifier()(expr); }
+  static Expr Simplify(Expr expr, String mode) { return NormSimplifier(mode)(expr); }
```

Comment:
You can check mode is in ['eval', 'training'] and report error in Simplify()

Cuz checking it in wrapper function is better than in `VisitBinding_` I think